### PR TITLE
Skip all tests when engines.atom has been bumped to exclude the current Atom version

### PR DESCRIPTION
**Pull request by smashwilson**
In #1512, I'm introducing a dependency on functionality that was added to Atom in atom/atom#17736, which is currently only available in nightly or dev builds of Atom. To capture this, I want to be able to bump the semver range specified in the `.engines.atom` field of our `package.json` to document the minimum supported Atom version, but our CI matrix rows for stable and beta will fail for _everything_ until that change makes it way to a stable Atom release.

This adds a check for compatibility between the declared Atom version range and the current host Atom range. If the Atom instance running the test suite is not expected to succeed because it's out of that range, we skip all tests. This will result in a "passing" test suite.

### CODE REVIEW

File: test/runner.js [hunk 1]
```diff
@@ -79,6 +80,24 @@ module.exports = createRunner({
...
+  const atomReleaseChannel = atomEnv.getReleaseChannel();
+  atomEnv.destroy();
+
+  if (!semver.satisfies(semver.coerce(atomVersion), engineRange)) {
+    process.stderr.write(
+      `Atom version ${atomVersion} does not satisfy the range "${engineRange}" specified in package.json.\n` +
+      `This version of atom/github is currently incompatible with the ${atomReleaseChannel} ` +
+      'Atom release channel.\n',
+    );
+    // A regular expression that matches nothing.
```
Comment (annthurium) [inline test/runner.js]: just for my own education: why do we need a regex that matches nothing?

Comment (smashwilson) [reply test/runner.js]: Well... I'm trying to find a graceful way to "cancel" the current test run and exit with a 0 status from the test runner. Mocha's `.grep()` method lets you filter the tests that execute by matching their titles against a regular expression, so providing it a regexp that will never match is equivalent to saying "don't run anything."

It's kind of awkward though. And I think it causes some misleading output from Mocha, although I tried to mitigate that with the `process.stderr` lines up above.

I tried doing a `process.exit(0)` first, but that caused the main process to think the renderer process had crashed and exit with a 1. Before that I was going to do this in a standalone script called from `script/cibuild`, but AppVeyor and CircleCI are both using the [atom/ci build scripts](https://github.com/atom/ci) which are hard-coded to run `atom --test test/`, so the test runner is the earliest common place to put the version check logic.

Now I'm thinking the Right Thing To Do :tm: is to do a little work upstream in atom-mocha-test-runner to support a clean early exit.

**Comment by smashwilson** [comment]
Here we go:

screen shot 2018-08-24 at 9 02 48 am

screen shot 2018-08-24 at 9 03 41 am

screen shot 2018-08-24 at 9 20 16 am

I didn't end up messing with the status or checks APIs to report "you can't deploy to this channel now" after all, though. It would take a moderate amount of work to put in place, we wouldn't be able to see it where we really need it (on the release branches when we're tagging a version!), and it'd be obsolete if we move into atom/atom eventually anyway.

We'll just need to remember which versions of atom/github rely on newer Atom features, by leaving a comment on the Atom ship-ready issue maybe? In any case it hasn't come up that often.

@annthurium what do you think?

**Comment by annthurium** [comment]
@smashwilson I'm totally okay with the lower-effort version of this - agree that it hasn't arisen often since usually we can programmatically check for new functionality in atom core.

**Pull request merged**